rusk/src/lib/ws.rs

```diff
@@ -76,122 +101,316 @@ async fn listening_loop<N: Network, DB: database::DB>(
+    mut shutdown: broadcast::Receiver<Infallible>,
+) where
+    VM: QueryRaw,
+    S: AsyncRead + AsyncWrite + Unpin,
+{
+    let (responder, mut responses) = mpsc::unbounded_channel::<Response>();
 
-impl Request {
-    fn parse(bytes: &[u8]) -> Result<Self, Box<dyn std::error::Error>> {
-        let (headers, bytes) = parse_header(bytes)?;
-        let (target_type, bytes) = parse_target_type(bytes)?;
-        let (target, bytes) = parse_string(bytes)?;
-        let (topic, bytes) = parse_string(bytes)?;
-        let data = bytes.to_vec();
-        Ok(Self {
-            headers,
-            target_type,
-            target,
-            topic,
-            data,
-        })
+    if let Ok(mut stream) = accept_async(stream).await {
+        loop {
+            tokio::select! {
+                // If the server shuts down we send a close frame to the client
+                // and stop.
+                _ = shutdown.recv() => {
+                    let _ = stream.close(Some(CloseFrame {
+                        code: CloseCode::Away,
+                        reason: Cow::from("Shutting down"),
+                    })).await;
+                    break;
+                }
+
+                rsp = responses.recv() => {
+                    // `responder` is never dropped so this can never be `None`
+                    let rsp = rsp.unwrap();
+
+                    // Serialize the response to text. If this does not succeed,
+                    // we simply serialize an error response.
+                    let rsp = serde_json::to_string(&rsp).unwrap_or_else(|err| {
+                        serde_json::to_string(&Response {
+                            request_id: rsp.request_id,
+                            data: vec![],
+                            error: Some(format!("Failed serializing response: {err}")),
+                        }).expect("serializing error response should succeed")
+                    });
+
+                    // If we error in sending the message we send a close frame
+                    // to the client and stop.
+                    if stream.send(Message::Text(rsp)).await.is_err() {
+                        let _ = stream.close(Some(CloseFrame {
+                            code: CloseCode::Error,
+                            reason: Cow::from("Failed sending response"),
+                        })).await;
+                        break;
+                    }
+                }
+
+                msg = stream.next() => {
+                    match msg {
+                        Some(msg) => match msg {
+                            Ok(msg) => match msg {
+                                // We received a request and should spawn a new tas to handle it.
+                                Message::Text(msg) => {
+                                    match serde_json::from_str(&msg) {
+                                        Ok(req) => {
+                                            task::spawn(handle_request(
```

Comment:
It's valid point to spawn a task per request. However considering we have a `self.inner.lock();`  in `vm.query_raw`  there is no much benefit of the concurrency here.